### PR TITLE
Apply initial draft of SunPy template

### PR DESCRIPTION
**Pull request by CyclingNinja**

### CODE REVIEW

File: {{ cookiecutter.package_name }}/.gitignore [hunk 1]
```diff
@@ -25,12 +29,7 @@ share/python-wheels/
 .installed.cfg
 *.egg
 MANIFEST
-{%- if cookiecutter.enable_dynamic_dev_versions == 'y' %}
-{{ cookiecutter.module_name }}/_version.py
-{% else %}
-{{ cookiecutter.module_name }}/version.py
-{% endif -%}
-
+sunpy/_version.py
```
Comment (Cadair) [inline {{ cookiecutter.package_name }}/.gitignore]: Revert

File: {{ cookiecutter.package_name }}/.gitignore [hunk 2]
```diff
@@ -145,21 +144,146 @@ venv.bak/
 
 # mypy
 .mypy_cache/
-.dmypy.json
-dmypy.json
 
-# Pyre type checker
-.pyre/
```
Comment (Cadair) [inline {{ cookiecutter.package_name }}/.gitignore]: Revert

File: {{ cookiecutter.package_name }}/.gitignore [hunk 3]
```diff
@@ -145,21 +144,146 @@ venv.bak/
...
+examples/**/*.asdf
+examples/**/*.csv
+figure_test_images*
+tags
+sunpy_map.asdf
+baseline
+docs/guide/data_types/figure.png
+october_M1_flares.csv
+aia_map.fits
+docs/tutorial/figure.png
```
Comment (Cadair) [inline {{ cookiecutter.package_name }}/.gitignore:271]:
```suggestion
```

File: {{ cookiecutter.package_name }}/.gitignore [hunk 4]
```diff
@@ -145,21 +144,146 @@ venv.bak/
...
-#  option (not recommended) you can uncomment the following to ignore the entire idea folder.
-#.idea/
+# asv stuff
+asv_env
+asv_results
+html
+results
+result_images
+# Save from Map.save example
+aia171.fits
```
Comment (Cadair) [inline {{ cookiecutter.package_name }}/.gitignore]: revert

Comment (nabobalis) [reply {{ cookiecutter.package_name }}/.gitignore]: Leave that aia file alone

File: {{ cookiecutter.package_name }}/docs/conf.py [hunk 5]
```diff
@@ -1,77 +1,366 @@
...
+    "drms": ("https://docs.sunpy.org/projects/drms/en/v0.6.4.post1/", None),
+    "hvpy": ("https://hvpy.readthedocs.io/en/latest/", None),
+    "matplotlib": ("https://matplotlib.org/stable", None),
+    "mpl_animators": ("https://docs.sunpy.org/projects/mpl-animators/en/stable/", None),
+    "pandas": ("https://pandas.pydata.org/pandas-docs/stable/", None),
+    "parfive": ("https://parfive.readthedocs.io/en/stable/", None),
+    "reproject": ("https://reproject.readthedocs.io/en/stable/", None),
+    "skimage": ("https://scikit-image.org/docs/stable/", None),
+    "spiceypy": ("https://spiceypy.readthedocs.io/en/stable/", None),
+    "sunkit_image": ("https://docs.sunpy.org/projects/sunkit-image/en/stable/", None),
```
Comment (nabobalis) [inline {{ cookiecutter.package_name }}/docs/conf.py:202]: Should sunpy be here?

File: {{ cookiecutter.package_name }}/docs/conf.py [hunk 6]
```diff
@@ -1,77 +1,366 @@
...
+    # Comes from the theme.
+    "default_thumb_file": PNG_ICON,
+    'abort_on_example_error': False,
+    'plot_gallery': 'True',
+    'remove_config_comments': True,
+    'doc_module': ('sunpy'),
+    'only_warn_on_example_error': True,
+}
+
+# -- Linking to OpenCV docs by using rst_epilog --------------------------------
```
Comment (nabobalis) [inline {{ cookiecutter.package_name }}/docs/conf.py:295]: Do we need this in the template?

File: {{ cookiecutter.package_name }}/{{ cookiecutter.module_name }}/.pre-commit-config.yml [hunk 7]
```diff
@@ -0,0 +1,8 @@
+exclude: "\\.asdf$"
+repos:
+  # This should be before any formatting hooks
+  - repo: https://github.com/astral-sh/ruff-pre-commit
+    rev: "v0.1.15"
+    hooks:
+      - id: ruff
+        args: ["--fix"]
```
Comment (nabobalis) [inline {{ cookiecutter.package_name }}/{{ cookiecutter.module_name }}/.pre-commit-config.yml]: Do we want to discuss ruff formatting here?

Comment (CyclingNinja) [reply {{ cookiecutter.package_name }}/{{ cookiecutter.module_name }}/.pre-commit-config.yml]: Sure, what do you have in mind specifically?

File: {{ cookiecutter.package_name }}/pyproject.toml [hunk 8]
```diff
@@ -5,7 +5,6 @@ requires = [
   "wheel",
 {%- if cookiecutter.use_compiled_extensions == 'y' %}
```
Comment (Cadair) [inline {{ cookiecutter.package_name }}/pyproject.toml:6]: This should look like this, though this change should be made in the openastronomy repo.
```suggestion
{%- if cookiecutter.use_compiled_extensions == 'y' %}
  "extension-helpers",
  "numpy>=1.25",
  "cython",
```

File: {{ cookiecutter.package_name }}/.gitignore [hunk 9]
```diff
@@ -145,21 +150,146 @@ venv.bak/
...
 # PyCharm
-#  JetBrains specific template is maintained in a separate JetBrains.gitignore that can
-#  be found at https://github.com/github/gitignore/blob/main/Global/JetBrains.gitignore
-#  and can be added to the global gitignore or merged into this file.  For a more nuclear
-#  option (not recommended) you can uncomment the following to ignore the entire idea folder.
-#.idea/
+./idea
+
+# Save from Map.save example
+aia171.fits
```
Comment (Cadair) [inline {{ cookiecutter.package_name }}/.gitignore:295]:
```suggestion
```

File: {{ cookiecutter.package_name }}/.gitignore [hunk 10]
```diff
@@ -145,21 +150,146 @@ venv.bak/
...
+# Log files generated by 'vagrant up'
+*.log
 
 # PyCharm
-#  JetBrains specific template is maintained in a separate JetBrains.gitignore that can
-#  be found at https://github.com/github/gitignore/blob/main/Global/JetBrains.gitignore
-#  and can be added to the global gitignore or merged into this file.  For a more nuclear
-#  option (not recommended) you can uncomment the following to ignore the entire idea folder.
-#.idea/
+./idea
```
Comment (nabobalis) [inline {{ cookiecutter.package_name }}/.gitignore:292]:
```suggestion
.idea
```
?

File: {{ cookiecutter.package_name }}/.gitignore (hunk 9 above)
Comment (nabobalis) [inline {{ cookiecutter.package_name }}/.gitignore:295]:
```suggestion
```

File: {{ cookiecutter.package_name }}/.gitignore (hunk 10 above)
Comment (nabobalis) [inline {{ cookiecutter.package_name }}/.gitignore:292]:
```suggestion
.idea
```

File: {{ cookiecutter.package_name }}/.gitignore [hunk 11]
```diff
@@ -145,21 +150,146 @@ venv.bak/
...
+tags
+sunpy_map.asdf
+baseline
+docs/guide/data_types/figure.png
+october_M1_flares.csv
+aia_map.fits
+docs/tutorial/figure.png
+
+### Pycharm
+.idea
```
Comment (nabobalis) [inline {{ cookiecutter.package_name }}/.gitignore:274]:
```suggestion
```

File: {{ cookiecutter.package_name }}/.pre-commit-config.yml [hunk 12]
```diff
@@ -0,0 +1,17 @@
...
+      - id: check-ast
+      - id: check-case-conflict
+      - id: trailing-whitespace
+        exclude: ".*(.fits|.fts|.fit|.header|.txt)$"
+      - id: check-yaml
+      - id: debug-statements
+      - id: check-added-large-files
+        args: [ "--enforce-all", "--maxkb=1054" ]
+      - id: end-of-file-fixer
+        exclude: ".*(.fits|.fts|.fit|.header|.txt|tca.*|.json)$|^CITATION.rst$"
```
Comment (nabobalis) [inline {{ cookiecutter.package_name }}/.pre-commit-config.yml:15]: I still think these type of files should be added to a global exclude.

File: {{ cookiecutter.package_name }}/.rtd-environment.yml [hunk 13]
```diff
@@ -0,0 +1,9 @@
+name: rtd_sunpy
+channels:
+  - conda-forge
+dependencies:
+  - python=3.10
+  - pip
+  - openjpeg
+  - graphviz!=2.42.*,!=2.43.*
```
Comment (nabobalis) [inline {{ cookiecutter.package_name }}/.rtd-environment.yml:8]: We should check if RTD's base image has updated to not have these versions. 

Then we can drop conda on RTD if so.

File: {{ cookiecutter.package_name }}/.rtd-environment.yml [hunk 14]
```diff
@@ -0,0 +1,9 @@
...
+  - pycairo
```
Comment (nabobalis) [inline {{ cookiecutter.package_name }}/.rtd-environment.yml:9]:
```suggestion
```
This surely is only for sunpy?

File: {{ cookiecutter.package_name }}/tox.ini [hunk 15]
```diff
@@ -1,63 +1,121 @@
...
 min_version = 4.0
 envlist =
-    py{38,39,310,311}-test
-    py38-test-oldestdeps
-    build_docs
-isolated_build = true
+    py{310,311,312}{,-oldestdeps,-devdeps,-online,-figure,-conda}
+    build_docs{,-gallery}
+    codestyle
+    base_deps
```
Comment (nabobalis) [inline {{ cookiecutter.package_name }}/tox.ini:7]: I guess this changes after a rebase?

File: {{ cookiecutter.package_name }}/tox.ini [hunk 16]
```diff
@@ -1,63 +1,121 @@
...
+description = Check the test suite does not fail if all optional dependencies are missing
 extras =
-    docs
+deps =
+    astropy
+    numpy
+    parfive[ftp]
 commands =
-    sphinx-build -j auto --color -W --keep-going -b html -d _build/.doctrees . _build/html {posargs}
+    python -c "import sunpy; sunpy.self_test()"
```
Comment (nabobalis) [inline {{ cookiecutter.package_name }}/tox.ini:121]: THis test env should go?

**Pull request closed without merging**